### ISSUE TITLE
ConvMolFeaturizer does not work with qm9

### ISSUE THREAD

**Issue by shayestehhamid** [deepchem/deepchem#2334]
## 🐛 Bug
Cannot generate features for QM9 with ConvMolFeaturizer. 

Additionally, this featurizer works with the example in the mol_graph_conv_featurizer.py. 

## To Reproduce
just test the following line:
dc.molnet.load_qm9(featurizer=dc.feat.ConvMolFeaturizer())

Here is my output: 

Failed to featurize datapoint 1663, [H]N(C(=O)N(C([H])([H])[H])C([H])([H])[H])C([H])([H])[H]. Appending empty array
Exception message: Pre-condition Violation
	getNumImplicitHs() called without preceding call to calcImplicitValence()
	Violation occurred on line 188 in file Code/GraphMol/Atom.cpp
	Failed Expression: d_implicitValence > -1
	RDKIT: 2020.09.3
	BOOST: 1_74

[08:17:56] 

****
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
Violation occurred on line 188 in file /home/conda/feedstock_root/build_artifacts/rdkit_1607861336837/work/Code/GraphMol/Atom.cpp
Failed Expression: d_implicitValence > -1
****
_[ This happens for all the samples ]_

## Expected behavior
the returned value should be the dataset featurized with no error. 

## Environment

* OS: Ubuntu 20.04.1
* Python version: 3.7.9
* DeepChem version: 2.4.0.dev
* RDKit version (optional): 2020.09.3
* TensorFlow version (optional): 2.3.1
* PyTorch version (optional): N/A
* BOOST: 1_74

## Additional context

I had no problem until the last update.

**Comment by rbharath** [comment]
Thanks for the report! I'm able to confirm the error. I'm not sure what could be causing it. I'm on RDKIT 2020.03.2 on my development machine. I don't think we changed anything in our handling of molecular handling recently.

Does anyone have any ideas on what's happening here?

**Comment by nd-02110114** [comment]
I can reproduce this error and I found the same error with CircularFingerprint. The reason this error happens is that the RDKit mol objects are not sanitized. Currently, SDFLoader loads unsanitized RDKit mol objects. I seem that this is because `gdb9.sdf` contains some molecules which can not be sanitized and we can not load all molecules if we set sanitize option as True.

https://github.com/deepchem/deepchem/blob/01ea79daeb1a02f2c7ce3592cfa2c0930ef883b7/deepchem/data/data_loader.py#L714-L731

And, this problem is similar to https://github.com/deepchem/deepchem/issues/1569.

Quick reproduce codes

```
>>> import rdkit
>>> rdkit.__version__
'2020.09.3'
>>> from rdkit import Chem
>>> import deepchem as dc
>>> smiles = ['[H]N(C(=O)N(C([H])([H])[H])C([H])([H])[H])C([H])([H])[H]']
>>> unsanitized_mols = [Chem.MolFromSmiles(val, sanitize=False) for val in smiles]
>>> conv_feat = dc.feat.ConvMolFeaturizer()
>>> conv_feat.featurize(unsanitized_mols)
[18:50:52]

****
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
Violation occurred on line 190 in file /Users/runner/miniforge3/conda-bld/rdkit_1607861350394/work/Code/GraphMol/Atom.cpp
Failed Expression: d_implicitValence > -1
****

Failed to featurize datapoint 0, [H]N(C(=O)N(C([H])([H])[H])C([H])([H])[H])C([H])([H])[H]. Appending empty array
Exception message: Pre-condition Violation
	getNumImplicitHs() called without preceding call to calcImplicitValence()
	Violation occurred on line 190 in file Code/GraphMol/Atom.cpp
	Failed Expression: d_implicitValence > -1
	RDKIT: 2020.09.3
	BOOST: 1_74

array([], shape=(1, 0), dtype=float64)
>>> ecfp = dc.feat.CircularFingerprint()
>>> ecfp.featurize(unsanitized_mols)
[18:51:34]

****
Pre-condition Violation
RingInfo not initialized
Violation occurred on line 45 in file /Users/runner/miniforge3/conda-bld/rdkit_1607861350394/work/Code/GraphMol/RingInfo.cpp
Failed Expression: df_init
****

Failed to featurize datapoint 0, [H]N(C(=O)N(C([H])([H])[H])C([H])([H])[H])C([H])([H])[H]. Appending empty array
Exception message: Pre-condition Violation
	RingInfo not initialized
	Violation occurred on line 45 in file Code/GraphMol/RingInfo.cpp
	Failed Expression: df_init
	RDKIT: 2020.09.3
	BOOST: 1_74

array([], shape=(1, 0), dtype=float64)
```

I think the easy fix is changing the sanitize option from False to True. 
On the other hand, if we change the option, we can not load all molecules (The size of dataset will change from 13K to 11K) and it is not appropriate to perform benchmark tests.

**Comment by rbharath** [comment]
Thank you for the analysis! I think it would be appropriate for us to change the flag to `sanitize=True`. As you point out, the same issue came up in #1569 and using sanitized molecules seems cleaner. We're working towards the MoleculeNet 2.0 release in any case so we can make a note of this modification in the released paper.

CC @mufeili in case he's working on QM9 for benchmarking

**Comment by peastman** [comment]
Do we know why sanitizing fails?  Is there a way we can make it work without throwing out a lot of molecules?

**Comment by rbharath** [comment]
I'm working on understanding more about this. When I turn sanitation on, I see some number of failures with messages of the type:

```
[12:39:12] Explicit valence for atom # 5 C, 5, is greater than permitted
[12:39:12] ERROR: Could not sanitize molecule ending on line 5729584
[12:39:12] ERROR: Explicit valence for atom # 5 C, 5, is greater than permitted
```

The original MoleculeNet paper lists QM9 as having 133885 compounds while the sanitized dataset has 132480. The difference all appears to be compounds like those listed above with misspecified valences. I think what was happening is that earlier versions of RDKit were more lax and allowed in compounds with misspecified valences while the newer versions don't allow for this. #1569 notes that the default of `sanitize=False` also turns off aromatic bonds for QM9 which doesn't sound right.

Given these facts, I think the sensible fix would be to turn on sanitization by `SDFLoader`. I'll put up a PR for review with this fix shortly.

**Comment by peastman** [comment]
Does this reflect errors in our file containing the dataset?  Are we specifying those molecules incorrectly?

**Comment by rbharath** [comment]
I believe that these errors reflect errors in the source SDF file containing the datasets yes. That is, the source specification itself has errors

**Comment by peastman** [comment]
Where did we get that file from?  Can we get a corrected version?

**Comment by rbharath** [comment]
The latest version of the QM9 dataset looks to be up on figshare https://figshare.com/collections/Quantum_chemistry_structures_and_properties_of_134_kilo_molecules/978904. I downloaded it and took a look. The tricky thing is that this dataset is just ~134K xyz files. I believe that RDKit doesn't natively support loading xyz files so we'd need to use an auxiliary tool like https://github.com/jensengroup/xyz2mol to convert the xyz files to SDF/CSV formats for us to be able to load.

I'm not sure if the latest version of the dataset has these issues corrected though. The release notes there note ~3K compounds which have some geometry issues (https://figshare.com/articles/dataset/Uncharacterized%3A_List_of_3054_molecules_which_failed_the_geometry_consistency_check/1057644?backTo=/collections/Quantum_chemistry_structures_and_properties_of_134_kilo_molecules/978904).

**Comment by rbharath** [comment]
I'm doing some more digging here. It turns out QM7 and QM8 are also failing to load with the same errors as QM9. Turning sanitization on for these loaders fixes the issue but at the cost of some additional compounds getting dropped off as well. In particular,  the new sanitized versions return the following numbers of compounds:

QM7:  6838/7160
QM8: 21747/21786

None of the other Molnet loaders use `SDFLoader` so I think these are our only datasets with issues.

**Comment by peastman** [comment]
It's probably reasonable as a workaround, but for the next release we should try to regenerate them from the original source and see if we can fix the problems properly.  This appears to be the authoritative source: https://springernature.figshare.com/collections/Quantum_chemistry_structures_and_properties_of_134_kilo_molecules/978904

**Comment by rbharath** [comment]
Yes agreed we should try to regenerate the datasets from the original source as part of the next release!

Let me go ahead and merge in #2351 as a workaround fix then so we can get the ball rolling on 2.4.0, but I'll keep this issue open for us to coordinate the needed fixes to QM9 for the next release for regenerating the source dataset.
